File: python/falsaposi.py
import math
epsilon = 0.0000001

def funcaoPoli(x):
    grau = 5
    coeficientes = [0.5,1.5,-2,-5,1,2]
    y = coeficientes[grau]
    for i in range(grau):
        y+=coeficientes[i]*x**(grau-i)
    
    return y

def funcaoTrig(x):
    y = 2-x- 2*math.sin(2*x)
    return y

def isolamento(limInf, limSup, fun):
    delta = 0.1
    minimo = limInf
    maximo = limSup

    funcao = fun
    anterior = funcao(minimo)
    intervalos = [];
    
    passo = minimo

    while(passo <= maximo):
        passo += delta
        y = funcao(passo)
        if(anterior*y < 0):
            intervalos.append(passo-delta)
            intervalos.append(passo)
        anterior = y
    
    return intervalos

def posFalsa(inf, sup, precisao, fun):
    a = inf
    b = sup
    x = 0
    funcao = fun

    while(True):
        x = (a*funcao(b) - b*funcao(a))/(funcao(b)-funcao(a))
        y = funcao(x)

        if(y*funcao(a) > 0):
            a = x
        else:
            b = x

        if((b-a) < precisao):
            x = (a+b)/2
            break
        elif(abs(y) < precisao):
            break
    
    return x

intervalos = isolamento(0, 3,funcaoTrig)#[0,3]-funcao trig/[-3.5,2.0] - funcaoPoli

i = 0

print("Raizes:\n")

while(i < len(intervalos)):
    raiz = posFalsa(intervalos[i], intervalos[i+1],epsilon, funcaoTrig)
    
    print(raiz)
    i+=2

File: cpp/falsaPosi.cpp
#include<iostream>
#include<iomanip>
#include<math.h>
#include<vector>

typedef long double ddouble;


ddouble polinomio(ddouble x){
    int grau =5;
    float coeficientes[grau+1] = {0.5,1.5,-2,-5,1,2};
    ddouble y = 0;
    for(int i = 0;i < grau; i++){
        y+=coeficientes[i]*powl(x,grau-i);
    };
    y+= (ddouble)coeficientes[grau];
    return y;
};

std::vector<ddouble> isolamento(){
    const float delta = 0.5f;
    float minimo = -3.5f;
    float maximo = 2.f;
    ddouble anterior = polinomio(minimo);
    std::vector<ddouble> intervalos;
    for(float i = minimo; i <=maximo; i+=delta){
        ddouble y = polinomio((ddouble)i);
        if(anterior*y < 0){
            intervalos.push_back(i-delta);
            intervalos.push_back(i);
        };
        anterior = y;
    };

    return intervalos;
};

ddouble falsaPosicao(ddouble inf, ddouble sup, ddouble precisao){
    ddouble a = inf;
    ddouble b = sup;
    ddouble x;
 
    while(true){
        x = (a*polinomio(b) - b*polinomio(a))/(polinomio(b) - polinomio(a));
        ddouble y = polinomio(x);

        if(y*polinomio(a) < 0){
            a = x;
        }else{
            b = x;
        };

        if(b-a < precisao){
            break;
        }else if(abs(y) < precisao){
            break;
        };
        

    };

    return(x);
};

int main(){
    ddouble epsilon = powl(2.L,-32.L);

    std::vector<ddouble> intervalos = isolamento();

    for(size_t i = 0; i <intervalos.size(); i+=2){
        ddouble raiz = falsaPosicao(intervalos[i],intervalos[i+1],epsilon);
        std::cout << std::setprecision(20) << raiz << '\n';
    };


    return 0;
};